### ISSUE TITLE
no-unused-prop-types sometimes fails to detect props in functions on stateless components

### ISSUE THREAD

**Issue by pfhayes** [yannickcr/eslint-plugin-react#885]
Here is a minimal example that triggers this:

``` jsx
const PagingBlock = function(props) {
  return (
    <span>
      <a onClick={() => props.previousPage()}/>
      <a onClick={() => props.nextPage()}/>
    </span>
 );
};

PagingBlock.propTypes = {
  nextPage: React.PropTypes.func.isRequired,
  previousPage: React.PropTypes.func.isRequired,
};
```

The error message is:

```
  18:17  error  'previousPage' PropType is defined but prop is never used  react/no-unused-prop-types
```

Notes:
- Converting this to a React.createClass with static propTypes fixes the issue
- Removing either `<a/>` tag produces the correct lint error, the false error only occurs when both are present
- The error is notified for the first tag (switching `nextPage` and `previousPage` inside the render function will make the error text say `nextPage` is defined but not used

**Comment by pfhayes** [comment]
cc @oliviakim95

**Comment by kevinSuttle** [comment]
Another test. 
```jsx
import React, { PropTypes } from 'react';
import './Button.css';
import Icon from '../Icons/Icon';
import cssClassPrefix from '../Styles/CSSClassPrefix';

const Button = props => (
  <button
    className={[`${cssClassPrefix}Button-${props.design}`]}
    type={props.type}
    disabled={props.disabled}
    onClick={() => props.onClick()}
    onFocus={() => props.onFocus()}
    onBlur={() => props.onBlur()}
    onMouseOver={() => props.onMouseOver()}
    onMouseOut={() => props.onMouseOut()}
  >
    {props.icon ? <Icon glyph={props.icon} /> : null }
    {props.value}
  </button>
  );

Button.displayName = 'Button';

Button.propTypes = {
  type: PropTypes.oneOf(['button', 'reset', 'submit']),
  design: PropTypes.oneOf(['filled', 'bordered', 'transparent']),
  icon: PropTypes.string,
  value: PropTypes.string,
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  onMouseOver: PropTypes.func,
  onMouseOut: PropTypes.func,
};

Button.defaultProps = {
  type: 'button',
  design: 'bordered',
  disabled: false,
}

export default Button;

```

Claims 4/5 of the `on*` function PropTypes aren't used, where the 5th is the last one.

**Comment by LiJinyao** [comment]
this  bug also triggers when use `...props`.
```
const PagingBlock = function(props) {
  return (
    <SomeChild {...props} />
 );
};

PagingBlock.propTypes = {
  nextPage: React.PropTypes.func.isRequired,
  previousPage: React.PropTypes.func.isRequired,
};
```
The error message is the same.

I think it should count all properties as used in this case.

**Comment by ljharb** [comment]
Agreed.

**Comment by tpai** [comment]
Same issue...

```
export default function Content( { post } ) {
    const { title, author, body } = post;
    return (
        <Panel>
            <Article
              title={ title }
              meta={ author }
            >
                <div dangerouslySetInnerHTML={ { __html: body } } />
            </Article>
        </Panel>
    );
}

Content.propTypes = {
    post: PropTypes.shape( {
        title: PropTypes.string.isRequired,
        author: PropTypes.string.isRequired,
        body: PropTypes.string.isRequired,
    } ),
};
```

```
'post.title' PropType is defined but prop is never used (react/no-unused-prop-types) [javascript/eslint]
'post.author' PropType is defined but prop is never used (react/no-unused-prop-types) [javascript/eslint]
'post.body' PropType is defined but prop is never used (react/no-unused-prop-types) [javascript/eslint]
```

**Comment by xkr47** [comment]
```js
const FormInput = (props) => {
  const {
    input,
    type = "text",
    ...extraProps
  } = props;
  return (
    <div id={input.name + "_view"}>
       <input type={type} id={input.name} {...input} {...extraProps}/>
    </div>
  );
};

FormInput.propTypes = {
  input: React.PropTypes.object.isRequired,
  required: React.PropTypes.bool,
  type: React.PropTypes.string
};
```

-> complains that `required` is unused even if it gets used through extraProps.

**Comment by kevinSuttle** [comment]
@xkr47 But it's not, though. How would Eslint know if it's not listed in the inline `defaultProps` literal or your return statement?

**Comment by ljharb** [comment]
I think because `extraProps` is a rest arg, every prop in it should be marked as used (and it can be statically known in this  case).

**Comment by nowke** [comment]
I have a similar problem with `PropTypes.shape`.

```js
MyComponent.propTypes = {
  error: PropTypes.shape({
    username: PropTypes.string,
    password: PropTypes.string,
  }),
}
```

And I pass `error` to the child component as a whole.  
`<MyChildComponent error={this.props.error} />`

ESLint throws 
```
error  'error.username' PropType is defined but prop is never used  react/no-unused-prop-types
error  'error.password' PropType is defined but prop is never used  react/no-unused-prop-types
```

**Comment by devalnor** [comment]
Probably same issue...

``` 
import React, { PropTypes } from 'react';

export default function Article(props) {
    const {
        homepageThumb: {
            url: imageUrl,
            width: imageWidth,
            height: imageHeight
        }
    } = props;

    return (

        <img
            src={imageUrl}
            width={imageWidth}
            height={imageHeight}
            alt="thumb"
        />

    );
}

Article.propTypes = {
    homepageThumb: PropTypes.shape({
        url: PropTypes.string,
        width: PropTypes.number,
        height: PropTypes.number
    }).isRequired
};
```
```
  59:20  error    'thumbnail_images.homepage-thumb.width' PropType is defined but prop is never used   react/no-unused-prop-types
  60:21  error    'thumbnail_images.homepage-thumb.height' PropType is defined but prop is never used  react/no-unused-prop-types
```

**Comment by BarryThePenguin** [comment]
I think I've figured out the source of this bug. Looking at @pfhayes example,

```js
const PagingBlock = function(props) {
  return (
    <span>
      <a onClick={() => props.previousPage()}/>
      <a onClick={() => props.nextPage()}/>
    </span>
 );
};

PagingBlock.propTypes = {
  nextPage: React.PropTypes.func.isRequired,
  previousPage: React.PropTypes.func.isRequired,
};
```

It looks like when [`utils.getParentStatelessComponent()`](https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/rules/no-unused-prop-types.js#L608) is called, the onClick `ArrowFunctionExpression` is returned instead of the PagingBlock function. In `utils.getParentStatelessComponent()` there is a regex test being performed on the [`Function` keyword](https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/util/Components.js#L347). This would return results for both `FunctionExpression` and `ArrowFunctionExpression`.

Need to find a way to differentiate between the two. I'll submit a PR if I can figure it out.

**Comment by iegik** [comment]
Still throws error when:

```
static propTypes = {
    data: React.PropTypes.arrayOf(React.PropTypes.any).isRequired,
};

static defaultProps = {
    data: [],
};

componentWillReceiveProps(nextProps) {
    this.state.data = nextProps.data; // Here ESLint do not see data prop
}
```

**Comment by ahauser31** [comment]
Yeah, for me usage of a prop only in `componentWillReceiveProps` also throws this linting error.

**Comment by BarryThePenguin** [comment]
You both may want to open another issue for `componentWillReceiveProps`

**Comment by borwahs** [comment]
@iegik  & @ahauser31 -

I ran into the same issue and found that if you destructure `nextProps` instead of using dot notation, the linting error is not thrown.

I also confirmed in this thread: https://github.com/yannickcr/eslint-plugin-react/issues/884#issuecomment-253903073

**Comment by geuis** [comment]
Still getting this in the most recent version of `eslint-plugin-react` in exactly the same code situation as the original example.

**Comment by BarryThePenguin** [comment]
@geuis there are [tests for the exact same code situation as the original example](https://github.com/yannickcr/eslint-plugin-react/blob/7ebcd486160277b4e13fed2a484156fc2e9420f6/tests/lib/rules/no-unused-prop-types.js#L1368-L1399).

If you have a different example that is still failing, would be worth opening a new issue

**Comment by wellingtonamaralleitao** [comment]
Also failing to detect props being used in react lifecycle fucntions like

``` javascript
...
modalClosed: PropTypes.bool.isRequired
```

``` javascript
componentWillReceiveProps(nextProps) {
    if (nextProps.modalClosed) {
      this.props.router.goBack()
    }
  }
```

**Comment by ljharb** [comment]
@wellingtonamaralleitao see #1232

**Comment by healqq** [comment]
I might do something wrong but this shows onItemClick propType not to be used:
```
import React from 'react';
import PropTypes from 'prop-types';

import './List.css';

function List(props) {
  return (
    <div className="List">
        I am a list
        {
          props.items.map(item =>
            (<div className="List-item">
              <button onClick={() => props.onItemClick(item.id)}>{item.title} {item.isAdded}
              </button>
            </div>),
          )
        }
    </div>
  );
}

List.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      title: PropTypes.string,
    }),
  ),
  onItemClick: PropTypes.func,
};
List.defaultProps = {
  items: [],
  onItemClick: null,
};
export default List;
```

**Comment by BarryThePenguin** [comment]
@healqq looks legit. I'd recommend opening a new issue

**Comment by TrevorSayre** [comment]
```
export default class Users extends Component {

  static propTypes = {
    ...
    onLoad: PropTypes.func.isRequired,
  };

  async componentWillMount() {
    await Promise.all([
      this.props.onLoad(),
      this.onSearch(),
    ]);
    ...
  }

  ...

}
```

`'onLoad' PropType is defined but prop is never used (react/no-unused-prop-types)`

**Comment by najinj** [comment]
Using eslint-plugin-react v^7.14.3 still throw this error

**Comment by chinhoe** [comment]
Experiencing the same issue too. Giving me an error saying 'PropType is defined but prop is never used' when my props is being used within a function inside a functional component.

**Comment by fiher** [comment]
I have the same error too, I am passing a prop to a functional component and I define defaultProps so I can make sure it is always a function (it is not required prop) and it marks it as unused

**Comment by prateekmaheshwari** [comment]
I have the same error too with eslint-plugin-react v^7.19.0

**Comment by GorgeousPuree** [comment]
How it was not fixed in 4 years??

**Comment by ljharb** [comment]
@GorgeousPuree the issue is closed, so it's fixed. if someone still has this problem on the latest version, please file a new issue and I'll be happy to take a look.

**Comment by GorgeousPuree** [comment]
@ljharb I'm sorry, pal, my bad. I've just mixed up this issue with another, which looks almost the same.

**Comment by shashwata27** [comment]
**`React.PropTypes`** is deprecated since 15.5.
I'm getting the same error, of unused but defined.
here's my code.

```
import PropTypes from "prop-types";

People.propTypes = {
  id: PropTypes.number,
  details: PropTypes.object,
  deleteFunc: PropTypes.func,
  importantFunc: PropTypes.func,
};
```

What am I doing wrong?

**Comment by ljharb** [comment]
@shashwata27 theres no component in that code. Please file a new issue if you’re still having trouble.

**Comment by shashwata27** [comment]
> @shashwata27 theres no component in that code. Please file a new issue if you’re still having trouble.

Check I updated the code above, should I make new issue?

**Comment by ljharb** [comment]
Yes please; commenting on old closed issues is not an efficient way to get help.
